ambry-router/src/test/java/com/github/ambry/router/CloudOperationTest.java

```diff
@@ -0,0 +1,881 @@
+      }
+    });
+    // Local data center, two nodes will return Disk_Unavailable, one node will return Blob_Not_Found
+    for (int i = 0; i < 3; ++i) {
+      if (i < 2) {
+        localDcServers.get(i).setServerErrorForAllRequests(ServerErrorCode.Disk_Unavailable);
+      } else {
+        localDcServers.get(i).setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+      }
+    }
+
+    GetBlobOptionsInternal options = new GetBlobOptionsInternal(new GetBlobOptionsBuilder().build(), false, routerMetrics.ageAtGet);
+    GetBlobOperation op = createGetBlobOperationAndComplete(blobId, null, options);
+
+    Assert.assertTrue("Operation should be complete at this time", op.isOperationComplete());
+    RouterException routerException = (RouterException) op.getOperationException();
+    if (routerException == null) {
+      Assert.fail("Expected getBlobOperation to fail");
+    }
+    Assert.assertEquals(RouterErrorCode.BlobDoesNotExist, routerException.getErrorCode());
+
+    mockServerLayout.getMockServers().forEach(MockServer::resetServerErrors);
+  }
+
+  /**
+   * Disk backed DC returns different kinds of errors while cloud backed DC returns Not Found.
+   */
+  @Test
+  public void testGetBlobSuccessInThePresenceOfVariousErrors() throws Exception {
+    int blobSize = 4096;
+    BlobProperties blobProperties = new BlobProperties(blobSize, "serviceId", "memberId", "contentType", false, Utils.Infinite_Time,
+        Utils.getRandomShort(random), Utils.getRandomShort(random), testEncryption, null, null, null);
+    byte[] userMetadata = new byte[10];
+    random.nextBytes(userMetadata);
+    byte[] putContent = new byte[blobSize];
+    random.nextBytes(putContent);
+    BlobId blobId = doPut(blobProperties, userMetadata, putContent);
+
+    // The put for the blob being requested happened.
+    String dcWherePutHappened = routerConfig.routerDatacenterName;
+
+    // GetBlobOperation to DC1 returns Not Found. Then will get blob from the DC3 where put happened.
+    Properties props = getDefaultNonBlockingRouterProperties(true, "DC1");
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+    doTestSuccessInThePresenceOfVariousErrors(blobId, dcWherePutHappened, blobSize, blobProperties, userMetadata, putContent);
+
+    // DC2 returns different errors. Then will get blob from the DC3 where put happened.
+    props = getDefaultNonBlockingRouterProperties(true, "DC2");
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+    doTestSuccessInThePresenceOfVariousErrors(blobId, dcWherePutHappened, blobSize, blobProperties, userMetadata, putContent);
+
+    // test requests coming in from local dc.
+    props = getDefaultNonBlockingRouterProperties(true, "DC3");
+    routerConfig = new RouterConfig(new VerifiableProperties(props));
+    doTestSuccessInThePresenceOfVariousErrors(blobId, dcWherePutHappened, blobSize, blobProperties, userMetadata, putContent);
```

Comment:
if we are including cloud data nodes in the test, then DC1 would be the cloud DC, and DC2,3,4 would the the disk based DC, shouldn't we test starting from DC2, not DC1, or you are also trying to test GetOperation from cloudDC?

Reply:
Yes, the test cases covered three sub test cases,
1. Start from Cloud DC which is DC1.
2. Start from the Ambry DC2 which is not the original DC put happened.
2. Start from the Ambry DC3 where put happened.